Add SSH module with persistent interactive shell support

sinkingsugar opened the pull request:
## Summary

- Implements new SSH module with 4 shards for remote shell execution over SSH
- Adds `SSH.Connect` for establishing persistent shell sessions with PTY support
- Adds `SSH.Execute` for running commands with automatic interactive detection
- Adds `SSH.SendInput` for handling interactive commands (python REPL, vim, etc.)
- Adds `SSH.Disconnect` for graceful connection cleanup
- Replaces HTTP-based shell server approach with direct SSH connections
- Maintains shell state (cd, export persist across commands)
- Includes automatic output cleaning (ANSI codes, command echo, prompts)

## Implementation Details

**Architecture:**
- Pure Rust implementation using russh and tokio
- Ref-counted `SSHShell` object for session management
- Timeout-based prompt detection for interactive command identification
- Shared tokio runtime for async SSH operations

**Shards:**
1. `SSH.Connect(Host: User: KeyPath: Port: Timeout:)` → SSHShell object
2. `SSH.Execute(Session: Timeout: CleanOutput:)` → Table with status and output
3. `SSH.SendInput(Session: Timeout:)` → Table with status and output  
4. `SSH.Disconnect` → None

**Features:**
- Persistent shell state (cd, export work across commands)
- Interactive command detection (no prompt within 3s timeout)
- Output cleaning (strips ANSI, command echo, prompts)
- SSH key and password authentication support
- PTY allocation for proper shell emulation
- Concurrent operation support via tokio runtime

## Test Plan

- [x] Basic command execution with cleaned output
- [x] Persistent state verification (cd, export across commands)
- [ ] Interactive command handling (python REPL, vim)
- [ ] Connection timeout and error handling
- [ ] Multiple concurrent SSH sessions
- [ ] Build verification on CI

## Related

This implementation replaces the HTTP server approach from `shell_server_cross_platform.py` with direct SSH, enabling remote shell execution without intermediate HTTP servers.

sinkingsugar committed:
Add rssh related shards to allow SSH sessions directly from shards

sinkingsugar committed:
fix test script for ssh

sinkingsugar committed:
try using ssh2 crate instead...

sinkingsugar committed:
Add SSH module with persistent shell and interactive command support

Implements native SSH connectivity for Shards to replace HTTP-based shell-cmd approach.

Features:
- SSH.Connect: Establishes persistent SSH shell sessions with PTY support
  - Hostname resolution (localhost, IPs, domains)
  - Password and SSH key authentication
  - Configurable connection timeout
- SSH.Execute: Executes commands in persistent shell
  - Automatic prompt detection with ANSI stripping
  - Interactive command detection via timeout
  - Clean output (removes ANSI codes, echo, prompts)
  - Persistent state (env vars, directory changes)
  - Returns status table with completion info
- SSH.SendInput: Sends input to interactive commands
  - Handles interactive sessions (read, prompts, etc.)
  - Waits for completion or next prompt
- SSH.Disconnect: Cleanly closes SSH session

Implementation:
- Uses ssh2 crate (libssh2 bindings) for synchronous SSH
- Links project's libressl instead of vendored OpenSSL
- Links system zlib for compression support
- Custom SSHShell object type with variable type support
- Comprehensive test suite with Docker SSH server

Files:
- shards/modules/ssh/: Complete Rust module implementation
- shards/tests/ssh-test.shs: Test script covering all features
- shards/tests/ssh-test-docker/: Docker-based SSH test server

sinkingsugar committed:
Use run_blocking for SSH operations and switch to trace logging

Changes:
- Implement BlockingShard trait for all SSH shards
- Use run_blocking() to run SSH operations in thread pool with coroutines
- Replace shlog_debug with shlog_trace for less verbose logging
- Fix lifetime annotations for activate_blocking (&'static str)

This integrates SSH operations properly with Shards' concurrency model,
preventing blocking on the main thread and enabling better async handling.

sinkingsugar committed:
Merge commit '11fdb6852ff1bbc2132e4bcce8090223cf518d34' into gio/ssh-shards

sinkingsugar committed:
Fix safety issues in SSH module and FS module

- SSH: Change mutable references to const for ref-counted SSHShell objects
- SSH: Handle mutex lock poisoning gracefully instead of panicking
- FS: Add integer overflow checking when converting int64 to size_t
- SSH: Disable module for Emscripten builds

sinkingsugar committed:
Remove integer overflow checks for chunkSize and maxSize in FS.Read, simplifying the code while maintaining functionality.

sinkingsugar committed:
Remove SSH.Disconnect shard and implement automatic cleanup in Drop

SSH sessions now automatically disconnect when dropped, following Shards'
stateful design where wires own resources. This simplifies the API by
removing the need for manual disconnect calls.

sinkingsugar committed:
Add connection failure detection and reporting to SSH module

Implement proper handling of network disconnections:
- Add is_connected state tracking to SSHShell
- Detect connection errors (broken pipe, connection reset, etc.)
- Return "connection_lost" status instead of misleading "requires_interaction"
- Fail early with clear errors on dead sessions
- Mark session as disconnected in Drop

sinkingsugar committed:
Add SSH.IsConnected shard for connection status checking

Add SSH.IsConnected shard that returns a boolean indicating whether
the SSH session is still connected. This enables proactive connection
checking for graceful error handling without relying on Maybe() blocks.

sinkingsugar committed:
build fixes

sinkingsugar committed:
Enhance SHMesh metadata handling and improve Set cleanup

- Updated SHMesh::setMetadata to include a 'force' parameter, allowing forced updates of metadata even if it already exists.
- Modified Set destructor and composeV2 method to ensure proper cleanup of table type information, preventing memory leaks.
- Adjusted metadata setting in Set to handle table types differently, ensuring correct behavior based on type changes.

sinkingsugar committed:
Add type validation for Rust shard parameters at compose time

This brings Rust shards to feature parity with C++ shards by validating
parameter types during composition using the battle-tested C++ TypeMatcher.

Implementation:
- Added shards_collect_required_variables_typed() C FFI export in runtime.cpp
- Added Rust FFI bindings and safe wrapper in util.rs
- Added has_context_variables() helper for runtime optimization
- Updated rust_macro to generate type-validated compose code with optimization

The validation now catches type mismatches at compose time with helpful errors:
  "No matching variable found for parameter Host, was: Var(Int),
   expected any of String Var(String)"

Previously, Rust shards only collected context variables without validating
their types against parameter declarations, allowing invalid types to pass
through compose and fail later at runtime.

sinkingsugar committed:
Improve exception handling in type validation FFI

Add intermediate catch for std::exception to provide better diagnostics
if unexpected errors occur during type validation. The three levels are:

1. ComposeError (expected) - Type validation failures
2. std::exception (unexpected) - Other errors with details
3. catch-all (very rare) - Non-exception errors

This helps distinguish between expected type mismatches and actual bugs.

sinkingsugar committed:
Fix unsafe pointer violation in type validation

Handle empty valid_types slices properly by using NonNull::dangling()
instead of as_ptr() which could return an invalid pointer.

This fixes a panic in compose:
  'slice::from_raw_parts_mut requires the pointer to be aligned and
   non-null, and the total size of the slice not to exceed isize::MAX'

The issue occurred when creating SHTypesInfo from an empty slice -
as_ptr() on an empty slice can return a dangling or null pointer,
violating the safety requirements for from_raw_parts_mut.

sinkingsugar committed:
Add null pointer checks in has_context_variables

Check that type info pointers are non-null before dereferencing
in has_context_variables(). Without these checks, accessing seq/table
type elements could crash if the pointer is null but len > 0.

This adds defensive checks for:
- seq_types.elements
- table_types.elements

Both are now verified non-null before iteration.

sinkingsugar committed:
Fix double-free in type validation by using temp C++ array

CRITICAL FIX: Prevent memory corruption from mixing C++ and Rust allocators.

The issue was passing a Rust Vec directly to C++ code that uses
shards::arrayPush() (C allocator). When C++ modified the array,
it allocated memory with C malloc, but Rust tried to free it with
its own allocator, causing double-free/invalid-free crashes.

Valgrind error:
  Invalid free() / delete / delete[] / realloc()
  Address 0x900000009 is not stack'd, malloc'd or (recently) free'd

Solution:
- Create temporary SHExposedTypesInfo with C++ allocator
- Let C++ populate it via shards::arrayPush()
- Copy results to Rust Vec
- Free C++ array with shards::arrayFree()
- Add shards_array_free() FFI function for proper cleanup

This ensures allocator consistency: C++-allocated memory freed by C++,
Rust-allocated memory freed by Rust.

sinkingsugar committed:
Fix UI.Window Position parameter to accept None

The Position parameter was declared with FLOAT2_VAR_SLICE which doesn't
allow None, but the code checks if position is None (line 241).

Changed to FLOAT2_VAR_OR_NONE_SLICE to match the actual behavior where
Position is optional.

This fixes the compose error:
  'No matching variable found for parameter Position, was: None,
   expected any of Float2 Var(Float2)'

sinkingsugar committed:
Fix Http.Stream to accept both object and variable reference

The Stream parameter only accepted Var(Http.Stream) but should also
accept the Http.Stream object directly.

When passing a variable like 'Http.Stream(picture-stream)', the type
system was interpreting it as Var(Var(Http.Stream)) instead of
recognizing the variable's value type.

Created STREAM_TYPE_OR_VAR to accept both:
- Http.Stream (direct object)
- Var(Http.Stream) (variable reference)

This fixes the compose error:
  'was: Var(Var(Http.Stream)), expected any of Var(Http.Stream)'

sinkingsugar committed:
Fix UI Queue parameter to accept direct object type

GFX_QUEUE_VAR_OR_NONE_TYPES only included Var(GFX.Queue) and None,
but should also accept the GFX.Queue object directly.

This matches the pattern from Http.Stream fix - parameters taking
custom objects need to accept both the object and variable references.

Changed to include:
- None
- GFX.Queue (direct object)
- Var(GFX.Queue) (variable reference)

sinkingsugar committed:
Fix UI.Area Position parameter to accept None

Position parameter was declared with [float2, float2_var] but should
also allow None since it's optional.

Added common_type::none to the accepted types.

sinkingsugar committed:
Fix egui optional parameters to accept None

Fixed parameters that were missing None in their type declarations
but actually handled None in their implementation:

AutoGrid:
- ColumnSpacing: FLOAT_VAR_SLICE -> FLOAT_VAR_OR_NONE_SLICE
- RowSpacing: FLOAT_VAR_SLICE -> FLOAT_VAR_OR_NONE_SLICE

Image & ImageButton:
- Scale: FLOAT2_VAR_SLICE -> FLOAT2_VAR_OR_NONE_SLICE
- Size: FLOAT2_VAR_SLICE -> FLOAT2_VAR_OR_NONE_SLICE

These parameters use unwrap_or or optional handling in their
activate functions, so they should declare None as valid.

sinkingsugar committed:
Add missing imports for _OR_NONE_SLICE constants

guusw commented on shards/rust/src/util.rs:
```diff
@@ -89,6 +90,83 @@ pub fn collect_required_variables(
+extern "C" {
+  fn shards_collect_required_variables_typed(
+    data: *const SHInstanceData,
+    out: *mut SHExposedTypesInfo,
```
This should go through SHCore

sinkingsugar replied:
does it matter? it's kinda very internal and rust is basically mandatory now due to cli etc anyway

guusw replied:
I mean if you ever write a DLL shard in rust you can not use them

guusw replied:
I'm slightly confused as to why bg3 shards don't fail to build though

sinkingsugar replied:
oh that's interesting, I didn't think about that..

sinkingsugar replied:
Something to test lol

sinkingsugar committed:
Fix Http.Stream parameter to use inline type array

Changed from lazy_static STREAM_TYPE_OR_VAR to inline array
[*STREAM_TYPE, *STREAM_TYPE_VAR] in shard_param macro. Also fixed
compose output type to use STREAM_TYPE instead of STREAM_TYPE_VAR.

sinkingsugar committed:
Fix UI.Area Pivot and Anchor parameters to accept None

Add common_type::none to Pivot and Anchor parameter types in both
AreaShard and SubAreaShard to allow None values, matching the pattern
used by other optional parameters like Position.

sinkingsugar committed:
Fix SSH shell prompt detection to distinguish interactive programs

Improve prompt detection logic to avoid false positives with interactive
programs like Python REPL. Exclude >>> and >> patterns which indicate
being inside a program rather than at the shell prompt.

Adjust timing and buffering:
- Add initial 200ms delay for shell to echo command
- Increase max wait iterations from 3s to 5s
- Require 2s of no-data before considering interactive
- Add final prompt check before declaring interactive

Add clarifying comments about connection state checking.

sinkingsugar committed:
Fix SSH command echo issues by switching to bash and improving output cleaning

This commit resolves multiple SSH terminal handling issues:

1. Switch to bash on connection: After initial shell connection, attempt to
   switch to bash using `exec bash --login`. This is necessary because zsh's
   ZLE (line editor) ignores `stty -icanon -echo` settings, causing command
   echoes to appear in output. Falls back gracefully if bash is unavailable.

2. Improved output cleaning: Added control character filtering to remove
   backspace (\b), carriage return (\r), and other control characters that
   ZSH ZLE injects for command-line redraw. This fixes issues where commands
   appeared with doubled first letters (ppwd, eecho, uuname).

3. Better Ctrl+C handling: After sending Ctrl+C to cancel interactive
   commands, properly drain the channel buffer to prevent leftover garbage
   from corrupting subsequent command outputs.

4. Added flush after command transmission: Ensures commands are fully
   transmitted before reading responses, preventing timing-related issues.

The implementation mirrors the working Python shell server approach while
adding proper error handling for systems without bash.

sinkingsugar committed:
Add macOS GPU test runner and various fixes

- Add parallel GPU test runner script with error detection and logging
- Fix hot-reload namespace parameter handling
- Add test-logs-* to gitignore
- Fix Tensor.Div to accept float variables

sinkingsugar committed:
Add test filtering options to macOS GPU test runner and fix AutoGrid optional parameters

sinkingsugar committed:
fix old Shards synthax reference

sinkingsugar committed:
Fix undefined behavior in shards_collect_required_variables_typed

Replaced dangerous reinterpret_cast between incompatible types (SHExposedTypesInfo* to ExposedInfo*) with a proper overload that works directly with SHExposedTypesInfo. This eliminates both undefined behavior and unnecessary copies.

- Added collectRequiredVariables overload that takes SHExposedTypesInfo& directly
- Removed reinterpret_cast, now passes correct type to the new overload
- Memory is properly managed at call site (verified in util.rs)

sinkingsugar committed:
Update Swift bindings

sinkingsugar committed:
Expose collectRequiredVariablesTyped through SHCore interface

Added SHCollectRequiredVariablesTyped to the SHCore interface to make
shards_collect_required_variables_typed available to DLL shards.

Changes:
- Added typedef and function pointer in shards.h
- Wired implementation in shardsInterface in runtime.cpp
- Updated Rust code to use Core interface instead of extern "C"
- Replaced shards_array_free calls with existing expTypesFree

sinkingsugar committed:
Add MaxOutputBytes parameter and output truncation to SSH.Execute

Added configurable output buffer size limit with tail-keeping truncation
to prevent memory issues with long-running commands.

Features:
- New MaxOutputBytes parameter (default: 65536, 0 = unlimited)
- Truncates from front, keeps tail when limit exceeded
- Adds "truncated": true to result table when truncation occurs
- Cleaned up unused imports
